File: frontend/src/gameBoard/svg.ts
import {h} from 'snabbdom'
import * as _ from 'lodash'

import {images} from '../common/images'
import {Scenario} from "../models";

const deltaLeft = 0.75
const deltaTop = Math.sqrt(3) / 2

export function left(x, y) { return deltaLeft * x }
export function top(x, y) { return  deltaTop * (y + ((x % 2 + 2) % 2) / 2) }
export function height(scenario) { return  (top(0, scenario.height) + 0.5) }
export function width(scenario) { return  (left(scenario.width, 0) + (1-deltaLeft)) }

export function directionToRotation(dir) {
  switch (Object.keys(dir)[0]) {
    case 'Up' : return 0
    case 'UpRight' : return 60
    case 'DownRight' : return 120
    case 'Down' : return 180
    case 'DownLeft' : return 240
    case 'UpLeft' : return 300
  }
}

export function translate(x, y){
  return `translate(${left(x, y)} ${top(x, y)})`
}

function imageAttrs(src, additionalAttrs?) {
  return _.merge({
    href: src,
    width: '1',
    height: '1',
    x: '0',
    y: '0',
  }, additionalAttrs)
}

export function tiles(scenario, tileClickListener) {
  return _.flatMap(_.range(0, scenario.width), x =>
      _.range(0, scenario.height)
          .map(y => h('image.tile', {
            attrs: imageAttrs(images.tile, {'data-x': x, 'data-y': y, transform: translate(x, y)}),
            style: {opacity: !scenario.pits.find(p => p.x === x && p.y === y) ? '1' : '0'},
            on: {click: tileClickListener}
          }))
  )
}

export function walls(scenario) {
  return scenario.walls.map(w => h('image', {
    attrs: imageAttrs(images.wall(w.direction), {
      transform: `${translate(w.position.x, w.position.y)}`
    })
  }))
}

export function target(scenario) {
  return h('image', {
    attrs: imageAttrs(images.target, {
      transform: translate(scenario.targetPosition.x, scenario.targetPosition.y)
    })
  })
}

export function traps(scenario) {
  return scenario.traps.map(function (trap) {
    if (trap.TurnRightTrap)
      return h('image', {
        attrs: imageAttrs(images.trapTurnRight, {
          transform: translate(trap.TurnRightTrap.position.x, trap.TurnRightTrap.position.y)
        })
      })
    else if (trap.TurnLeftTrap)
      return h('image', {
        attrs: imageAttrs(images.trapTurnLeft, {
          transform: translate(trap.TurnLeftTrap.position.x, trap.TurnLeftTrap.position.y)
        })
      })
    else if (trap.StunTrap)
      return h('use', {
        attrs: imageAttrs(images.trapStun, {
          transform: translate(trap.StunTrap.position.x, trap.StunTrap.position.y)
        })
      })
  })
}

export function startingPoints(scenario: Scenario) {
  return scenario.initialRobots.map((robot, index) =>
      h('image', {
        attrs: imageAttrs(images.playerStart(index), {
          transform: `${translate(robot.position.x, robot.position.y)} rotate(${directionToRotation(robot.direction)})`
        })
      })
  )
}

export function robots(game) {
  return game.robots.map(robot =>
      h('g', {attrs: {transform: 'translate(0.5 0.5)'}},
          h(`g#robot-translation-${robot.index}`, {attrs: {transform: translate(robot.position.x, robot.position.y)}},
              h(`g#robot-rotation-${robot.index}`, {attrs: {transform: `rotate(${directionToRotation(robot.direction)})`}},
                  h(`g#robot-scale-${robot.index}`, {attrs: {transform: 'scale(1)'}},
                      h('g', {attrs: {transform: 'translate(-0.5 -0.5)'}},
                          h('image', {attrs: imageAttrs(images.player(robot.index))})
                      )
                  ))
          )
      )
  )
}

File: frontend/src/frontend/gameBoard/Svg.scala
package frontend.gameBoard

import com.raquo.snabbdom.simple.implicits._
import com.raquo.snabbdom.simple.{VNode, attrs, tags}
import com.raquo.snabbdom.simple.styles.visibility
import com.raquo.snabbdom.simple.events.onClick
import frontend.common.Images
import frontend.util.Ui
import gameEntities._
import com.raquo.snabbdom.simple.attrs.id
import org.scalajs.dom.raw.{HTMLElement, MouseEvent}

object Svg extends Ui {
  val deltaLeft = 0.75
  val deltaTop = Math.sqrt(3) / 2


  def left(pos: Position): Double = deltaLeft * pos.x

  def top(pos: Position): Double = deltaTop * (pos.y + ((pos.x % 2d) + 2d) % 2d / 2d)

  def height(scenario: Scenario): Double = top(Position(0, scenario.height)) + 0.5

  def width(scenario: Scenario): Double = left(Position(scenario.width, 0)) + (1 - deltaLeft)

  def directionToRotation(dir: Direction) = dir match {
    case Up        => 0
    case UpRight   => 60
    case DownRight => 120
    case Down      => 180
    case DownLeft  => 240
    case UpLeft    => 300
  }

  def translate(pos: Position) =
    s"translate(${left(pos)} ${top(pos)})"

  def target(scenario: Scenario): VNode =
    tags.build("image")(
      attrs.build[String]("href") := Images.target,
      attrs.build[String]("width") := "1",
      attrs.build[String]("height") := "1",
      attrs.build[String]("x") := left(scenario.targetPosition).toString,
      attrs.build[String]("y") := top(scenario.targetPosition).toString,
    )

  def tiles(scenario: Scenario, click: Option[(Position, Direction) => Unit]): Seq[VNode] =
    for {
      x <- 0 until scenario.width
      y <- 0 until scenario.height
      p = Position(x, y)
    } yield tags.build("image")(
      attrs.build[String]("href") := Images.tile,
      attrs.build[String]("width") := "1",
      attrs.build[String]("height") := "1",
      attrs.build[String]("x") := left(p).toString,
      attrs.build[String]("y") := top(p).toString,
      if (scenario.pits.contains(p)) visibility := "hidden" else None,
      click.map(f => onClick := (event => f(p, event2direction(event))))
    )

  private def event2direction(event: MouseEvent): Direction = {
    val bb = event.target.asInstanceOf[HTMLElement].getBoundingClientRect()
    val dx = event.pageX - bb.left - bb.width / 2
    val dy = event.pageY - bb.top - bb.height / 2
    Math.floor((Math.atan2(dy, dx) / Math.PI * 3 + 6) % 6) match {
      case 0 => DownRight
      case 1 => Down
      case 2 => DownLeft
      case 3 => UpLeft
      case 4 => Up
      case 5 => UpRight
    }
  }

  def walls(scenario: Scenario): Seq[VNode] =
    scenario.walls.map(wall =>
      tags.build("image")(
        attrs.build[String]("href") := Images.wall(wall.direction),
        attrs.build[String]("width") := "1",
        attrs.build[String]("height") := "1",
        attrs.build[String]("x") := left(wall.position).toString,
        attrs.build[String]("y") := top(wall.position).toString,
      )
    )

  def traps(scenario: Scenario): Seq[VNode] =
    scenario.traps.map(trap =>
      tags.build("image")(
        attrs.build[String]("href") := Images.trap(trap),
        attrs.build[String]("width") := "1",
        attrs.build[String]("height") := "1",
        attrs.build[String]("x") := left(trap.position).toString,
        attrs.build[String]("y") := top(trap.position).toString,
      )
    )

  def startPoints(scenario: Scenario): Seq[VNode] =
    scenario.initialRobots.map(robot =>
      tags.build("image")(
        attrs.build[String]("href") := Images.playerStart(robot.index),
        attrs.build[String]("transform") := translate(robot.position)
          + " translate(0.5 0.5) "
          + s"rotate(${directionToRotation(robot.direction)})"
          + " translate(-0.5 -0.5)",
        attrs.build[String]("width") := "1",
        attrs.build[String]("height") := "1",
        attrs.build[String]("x") := "0",
        attrs.build[String]("y") := "0",
      )
    )

  def robots(robots: Seq[Robot]): Seq[VNode] = {
    robots.map(robot =>
      tags.build("g")(
        attrs.build[String]("transform") := "translate(0.5 0.5)",
        tags.build("g")(
          id := s"robot-translation-${robot.index}",
          attrs.build[String]("transform") := translate(robot.position),
          tags.build("g")(
            id := s"robot-rotation-${robot.index}",
            attrs.build[String]("transform") := s"rotate(${directionToRotation(robot.direction)})",
            tags.build("g")(
              id := s"robot-scale-${robot.index}",
              attrs.build[String]("transform") := s"scale(1)",
              tags.build("g")(
                attrs.build[String]("transform") := "translate(-0.5 -0.5)",
                tags.build("image")(
                  attrs.build[String]("href") := Images.player(robot.index),
                  attrs.build[String]("width") := "1",
                  attrs.build[String]("height") := "1",
                  attrs.build[String]("x") := "0",
                  attrs.build[String]("y") := "0",
                )
              )
            )
          )
        )
      )
    )
  }
}
